### PR TITLE
Implement worksheets API

### LINKED ISSUE #313: FileServer is not thread-safe
The FileServer is an asynchronous server that uses the ThreadingMixIn. Each request is executed in a newly created thread. The request handler methods in the FileServer access file_paths, file_handles and other variables without grabbing any lock. Thus, it's possible for modifications of these dictionaries to lead to inconsistent results.

### LINKED ISSUE #513: Remove BundleClients and compatibility hacks
Once REST API migration is complete, delete `RemoteBundleClient`, `LocalBundleClient`, as well as their lone dependencies and hacks introduced to maintain backwards compatibility.

Do a grep for TODOs.

### LINKED ISSUE #516: Remove support for connecting to 'local' instance
No longer be supported by the REST API.
Typical users should always connect to a server.
Advanced users can run their own servers.

### LINKED ISSUE #579: Worksheets REST API
Create REST API for worksheets, items, and permissions.

CLI expects certain interpreted format for worksheet JSON, opt for temporary "legacy" API that spits out this interpreted format for direct consumption by CLI just so that we can migrate to REST-only system asap.

### PR DESCRIPTION
**Pull request by kashizui**
Resolves #579 
Resolves #513 
Resolves #516 
Resolves #313

Sorry about the huge PR, I kind of went hog wild.

Cleaned up XMLRPC and BundleClients stuff as well.

Should be merged in after codalab/codalab-worksheets#219 because I ripped out the old bundle upload API.
- [x] ls
- [x] wls
- [x] work
- [x] finish search
- [x] finish mimic
- [x] new
- [x] wrm
- [x] wedit
- [x] detach
- [x] wperm
- [x] print
- [x] add
- [x] wadd
- [x] rebase
- [x] move remaining required LocalBundleClient functionality into legacy.BundleService.
- [x] check and debug test-cli locally
- [x] check and debug test-cli on circleci
- [x] test and debug the website
- [x] wizard to migrate users' default aliases
- [x] cleanup uuid resolution code (merge into canonicalize?)
- [x] fix completers

@percyliang (@yashsavani @ChengshuLi)

### CODE REVIEW

File: codalab/lib/bundle_cli.py
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Isn't `tuple_from_item` more accurate?

File: codalab/lib/cli_util.py
Comment (percyliang) [inline codalab/lib/cli_util.py]: given that this lives in `cli_util.py`, the name of this function should probably be a bit more specific (contain nested_dict or something)

File: codalab/lib/cli_util.py
Comment (percyliang) [inline codalab/lib/cli_util.py]: what's deep_get?

File: codalab/client/local_bundle_client.py
Comment (percyliang) [inline codalab/client/local_bundle_client.py]: Since we've gotten rid of the the XMLRPC server, why can't we get rid of the bundle client now?

File: codalab/lib/worksheet_util.py
Comment (percyliang) [inline codalab/lib/worksheet_util.py]: Hm, this makes me a bit uncomfortable.  Why do the client and server have different ways to resolve uuids?  Is the problem that in the server case, we want to make another REST API call to oneself?

Another related point is that even if the implementations were different, the functionality also seems different...it'd be nice if the functionality were the same and this was just an optimization (which we can explicitly test against).  This would make it easier to reason about.

Also, can we take this opportunity to create a new Python file that contains all the resolving logic and merge that with canonicalize.py?  The path from resolving strings to uuid starting with the CLI to the DB has been one of the most convoluted parts of the system...

**Review by percyliang — COMMENTED** [review]
Wow, good progress!

**Comment by percyliang** [comment]
Tried to run a job locally, and the worker seemed to have failed:

```
Traceback (most recent call last):
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 410, in _finish
    lambda: self._bundle_service.update_bundle_contents(
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 434, in _execute_bundle_service_command_with_retry
    f()
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 412, in <lambda>
    update_status))
  File "/home/pliang/research/codalab/codalab-cli/worker/bundle_service_client.py", line 21, in wrapper
    return f(*args, **kwargs)
  File "/home/pliang/research/codalab/codalab-cli/worker/bundle_service_client.py", line 150, in update_bundle_contents
    progress_callback=progress_callback)
  File "/home/pliang/research/codalab/codalab-cli/worker/rest_client.py", line 105, in _upload_with_chunked_encoding
    StringIO(response.read()))
BundleServiceException: Unable to update bundle contents in bundle service: Bad Request - remove got non-existent path:: /home/pliang/research/codalab/home/partitions/default/bundles/0xb362e13929e84069bd1786ab3d0750907r65UtZTtuC0
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 188, in _start
    self._monitor()
  File "/home/pliang/research/codalab/codalab-cli/worker/run.py", line 215, in _monitor
    self._worker.finish_run(self._uuid)
  File "/home/pliang/research/codalab/codalab-cli/worker/worker.py", line 228, in finish_run
    del self._runs[uuid]
KeyError: u'0xb362e13929e84069bd1786ab3d075090'
```

File: codalab/client/local_bundle_client.py
Comment (kashizui) [reply codalab/client/local_bundle_client.py]: I'm working on that :)
Currently, code in `legacy.BundleService` still uses it, but I'm looking at how much work it would be to just move all of that into `BundleService` since it's the only user.

**Comment by kashizui** [comment]
@percyliang What command did you run in your job?

**Comment by percyliang** [comment]
I just did 'run date' - that doesn't happen for you?

**Comment by kashizui** [comment]
It seems to work for me, but this is concerning.

**Comment by percyliang** [comment]
Can we try it out on wstest?

**Comment by percyliang** [comment]
Never mind, this was my problem - I was running two bundle managers in
parallel by accident.  It works now.

File: codalab/lib/worksheet_util.py
Comment (kashizui) [reply codalab/lib/worksheet_util.py]: @percyliang In the server case, we want to just call the model methods directly. In the client case, we have to make calls to the REST API.

I'll take a shot at cleaning this up!

Comment (kashizui) [reply codalab/lib/worksheet_util.py]: @percyliang by merging with canonicalize.py, you mean we don't actually want to "create a new Python file"?

Comment (percyliang) [reply codalab/lib/worksheet_util.py]: We do want to create a new file perhaps called `worksheet_resolver.py`.  We can move `canonicalize.py` into this class and delete `canonicalize.py`.

Comment (kashizui) [reply codalab/lib/worksheet_util.py]: I dramatically simplified the logic. I decided to keep canonicalize, since it's not a bad name and I was able to reduce all calls on the server side to canonicalize calls.

On the client side, I just added a few additional methods in BundleCLI that make the appropriate REST API calls to resolve uuids.

**Comment by kashizui** [comment]
@percyliang Should be ready for review!

All the *BundleClients, and a lot of other cruft is gone.
Given the amount of interlocking parts, I'm not sure how I could break this PR into more digestible chunks.. perhaps we could go over this some time in person.

File: codalab/lib/bundle_cli.py
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: reorder

File: codalab/lib/bundle_cli.py
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: simplify?

File: codalab/lib/bundle_cli.py
Comment (percyliang) [inline codalab/lib/bundle_cli.py:1067]: Shouldn't happen?

File: codalab/lib/bundle_cli.py
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: name always exists

File: codalab/lib/bundle_cli.py
Comment (kashizui) [inline codalab/lib/bundle_cli.py]: Use BundleCLI.simple_user_str

**Review by percyliang — APPROVED** [review]
LGTM modulo lots of further testing

**Comment by kashizui** [comment]
Addressed comments from this morning and did a bit more cleanup, merging as soon as CircleCI approves. This will temporarily disable website upload functionality on develop/staging, until we merge codalab/codalab-worksheets#219 in too.

File: codalab/lib/bundle_cli.py
Comment (kashizui) [reply codalab/lib/bundle_cli.py:1067]: So it turns out that this actually does happen, in the case where the bundle's contents have been deleted (using `cl rm -d`), `test-cli.py` has a check specifically for this case.

**Pull request merged**
